### DISCUSSION TITLE
time based Enhancements

### DISCUSSION THREAD

**Discussion by fast-turtle-9** [jertel/elastalert2#1477] [Q&A] [answered]
So I work with integrations, they have a connection status field. Elastic alert checks the statis every 15 mins. I use elastalert to check if the integration has failed or not by checking the integration status field for an "F". The problem is that, alot of times the integration fails for 2 cycles(30 mins) due to connection issues so I would like elastalert to alert me if the integration is on F status for three cycles. How would I make an enhancement as such, this is more of a question on the programming logic....any helpwould be appreciated....thanks

**Comment by jertel** [comment]
Did you consider using a frequency rule, that only alerts when the number of status=F records is > 2 in a 30 minute window?

**Reply by fast-turtle-9** [reply]
Yes, that actually could work, however there are quite a few integrations that are consistently F and would show up as false alerts. Due to those permanently F integrations I am using the change rule type. I was wondering if there is a variable or key that I could refer to as the count of that field in enhancements?

**Reply by fast-turtle-9** [reply]
I was working on some code, how would I set a while loop based off the data generated in the query, if I could count the number of Fs in the data that is queried? How are the documents stored when a query is run and how can they be called?

**Reply by jertel** [reply, marked as answer]
I dount think the count is currently saved in memory, so it's likely unavailable to the enhancement. But if you are _not_ using `use_count_query: true` then the `matches` list var inside of the enhancement should have the returned records from the query.

**Reply by fast-turtle-9** [reply]
okay, How would I be able to list the var from showing the data visually? if they are three seperate documents could I access them in enhancements, all three? and count them? 
I know so many questions, I really appreciate your help thank you!

**Reply by jertel** [reply]
I mentioned above that the records are available in the `matches` variable inside of the enhancement. So I don't understand your follow-up question. Am I missing something?

**Reply by fast-turtle-9** [reply]
My bad I understand now, I can basically iterate over the contents of the matches variable and see if there is a string match for integration_status: F and, +=1 a counter every time there is a match. thank you

**Reply by fast-turtle-9** [reply]
does the matches variable contain the (for example) 3 documents as 3 dictionaries within a list or is it simply one list? @jertel 

I am assuming its a dictionary in list because thats what makes sense would separate the data but just want to get your confirmation

**Reply by jertel** [reply]
It is a single match record (Dict of field/values) passed into the enhancement's `process()` method, not a list of Dicts like I had incorrectly remembered. However, the count of matches is pushed into each match record so you can read any of the process() invocations and find out how many total records matched:

```
from elastalert.enhancements import BaseEnhancement

class MyEnhancement(BaseEnhancement):

    # The enhancement is run against every match
    # The match is passed to the process function where it can be modified in any way
    # ElastAlert 2 will do this for each enhancement linked to a rule
    def process(self, match):
        total_matches = match["num_matches"]
```

Or, if you need to count the total number of field values that equal 'F' across those matched you could use a variable stored on the rule:

```
from elastalert.enhancements import BaseEnhancement

class MyEnhancement(BaseEnhancement):

    # The enhancement is run against every match
    # The match is passed to the process function where it can be modified in any way
    # ElastAlert 2 will do this for each enhancement linked to a rule
    def process(self, match):
        f_count = self.rule["accum_f_count"]
        if f_count is None:
            f_count = 0
    
        if match["custom_field"] == "F":
            f_count += 1

        self.rule["accum_f_count"] = f_count
```

Though I would think you will want to reset that count at some point but that's something you can figure out.

**Reply by fast-turtle-9** [reply]
this is exactly what I needed thanks so much, I will work on it from here!

**Reply by fast-turtle-9** [reply]
@jertel so I tried that but it seems I am missing something. 
let me bother you a bit more for how elastalert exactly works. 
`run_every:
  minutes: 10
type: change
timeframe:
  minutes: 65
this is part of my rule file, with this query timeframe how would elastalert pick up a match and query data?
say the data in elastic search gets updated every 15 minutes, so there would be 4 documents per query key for that query that would pass through the enhancement if there was a match, i would assume?
say first cycle was C second was F and third was F and the fourth was an F. would all three "F" documents be passed to the process function as  a match? or would it just be the first change in status from C to F?

**Reply by jertel** [reply]
It should only be the record that triggered the change. So when C goes to F I would expect to see the enhancement invoked with that specific F record. And when it goes back to C I would expect to see the enhancement invoked with the C record. I am going from memory so your best path forward is to use debugging to see what is passed into the enhancement in these scenarios.

**Reply by fast-turtle-9** [reply]
yeah based off what i figured out its not possible to do what i want with the change type rule only.
i would need to either have a mix of frequency and change or cardinality and change.
is it possible to have two rule types in one rule? or would I have to create a new rule? @jertel

**Reply by jertel** [reply]
You cannot mix two rule types into a single rule.

**Reply by fast-turtle-9** [reply]
okay thanks a lot for putting up with my questions, I think I will try to adjoin parts of the code of "change" rule type and the "frequency" rule type and see if I can't figure it out!
